internal/levee.go

```diff
@@ -39,75 +41,210 @@ func run(pass *analysis.Pass) (interface{}, error) {
 	if err != nil {
 		return nil, err
 	}
-	// TODO: respect configuration scope
 
+	// TODO: respect configuration scope
 	sourcesMap := pass.ResultOf[source.Analyzer].(source.ResultType)
 
-	// Only examine functions that have sources
-	for fn, sources := range sourcesMap {
-		for _, b := range fn.Blocks {
-			if b == fn.Recover {
-				// TODO Handle calls to sinks in a recovery block.
-				continue // skipping Recover since it does not have instructions, rather a single block.
+	for fn := range sourcesMap {
+		analyzeFn(fn, conf, pass)
+	}
+
+	return nil, nil
+}
+
+func analyzeFn(fn *ssa.Function, conf *config.Config, pass *analysis.Pass) {
+	graph := newGraph()
+
+	for _, p := range fn.Params {
+		if conf.IsSource(utils.Dereference(p.Type())) {
+			graph.addSource(p.Name(), p)
+		}
+	}
+
+	// this is needed to capture variables referred to by closures
+	for _, p := range fn.FreeVars {
+		switch t := p.Type().(type) {
+		case *types.Pointer:
+			if s, ok := utils.Dereference(t).(*types.Named); ok && conf.IsSource(s) {
+				graph.addSource(p.Name(), p)
 			}
+		}
+	}
 
-			for _, instr := range b.Instrs {
-				v, ok := instr.(*ssa.Call)
-				if !ok {
-					continue
+	for _, b := range fn.Blocks {
+		for _, inst := range b.Instrs {
+			value, _ := inst.(ssa.Value)
+
+			switch t := inst.(type) {
+			case *ssa.Alloc:
+				if conf.IsSource(utils.Dereference(t.Type())) && !source.IsProducedBySanitizer(t, conf) {
+					graph.addSource(value.Name(), t)
+				}
+
```

Comment:
This work was already done by the `source` analyzer, which should remain the source of truth for Source identification.